### ISSUE TITLE
Is it possible to deploy tkinter with tensorflow with Pyinstaller or Cx freeze?

### ISSUE THREAD

**Issue by Lucysmith8** [tensorflow/tensorflow#40485]
I've been trying to develop a Tkinter application with TensorFlow for two weeks now, But I'm having issues converting the application from .py to .exe file with both pyinstaller cx_freeze. 
I always get errors. Is it possible to develop a model built with keras without going through TensorFlow? If I move the TensorFlow library, can I then achieve what I am trying?

My app code :

from tensorflow.keras.models import load_model
from tensorflow.keras.preprocessing import image

import tkinter

def printVariable(event=0):
    #This function prints the option selected
    print (doc.get())

window = tkinter.Tk()

window.title("Combo Box Example")

lbldoctype = tkinter.Label(window, text="   Option Choice:    ")
lbldoctype.pack()

doc = tkinter.StringVar(window)
doc.set("Option 1")

combo = tkinter.OptionMenu(window, doc, "Option 1", "Option 2", "Option 3", "Option 4")
combo.pack()

btn = tkinter.Button(window, text="     Select     ", command=printVariable)
#pack the button into the window
btn.pack()

window.bind('<Return>', printVariable)

window.mainloop()

My setup code:

import cx_Freeze as cx

cx.setup(
    name='Taxon',
    version='1.0',
    description='retrain inception with a GUI',
    executables=[
        cx.Executable('myapp.py',
                      targetName='hephaistos', icon='myapp.ico')],
    options={
        'build_exe': {
            'packages': ['tensorflow'],
            'includes': ['idna.idnadata', 'zlib'],
            'excludes': ['PyQt4', 'PyQt5', 'PySide', 'IPython',
                         'jupyter_client', 'jupyter_core', 'ipykernel',
                         'ipython_genutils', 'Tkinter', 'matplotlib', 'h5py']

        }
    }
)

**Comment by ratansingh98** [comment]
Can you please provide description of your error?

**Comment by Lucysmith8** [comment]
Here is the error I get

![error](https://user-images.githubusercontent.com/66785574/84716926-c25a8200-af74-11ea-9706-72d85a87b5ff.jpg)

**Comment by Lucysmith8** [comment]
> Can you please provide a description of your error?

What I don't understand is that it searches for tensorflow in sites-application while I deposit it in the application folder

**Comment by ratansingh98** [comment]
> > Can you please provide a description of your error?
> 
> What I don't understand is that it searches for tensorflow in sites-application while I deposit it in the application folder

Have you used os.path.join in your codebase instead of a file path?

**Comment by Lucysmith8** [comment]
> Can you please provide a description of your error?
> 
> What I don't understand is that it searches for tensorflow in sites-application while I deposit it in the application folder
> 
> Have you used os.path.join in your codebase instead of a file path?

Yes, my code :

from os import path
import sys

if getattr(sys, 'frozen', False):
    MODEL_DIRECTORY = path.join(path.dirname(sys.executable), 'model')
else:
    MODEL_DIRECTORY = path.dirname(__file__)
    
    
mymodel = path.join(MODEL_DIRECTORY, 'mymodel.h5')

**Comment by Lucysmith8** [comment]
> Can you please provide a description of your error?
> 
> What I don't understand is that it searches for tensorflow in sites-application while I deposit it in the application folder
> 
> Have you used os.path.join in your codebase instead of a file path?

The problem may be in my setup? Because even without the code I sent, I still have the same problem. The problem can be found in this code?

    PYTHON_DIR = os.path.dirname (os.path.dirname (os .__ file__))
    os.environ ['TCL_LIBRARY'] = os.path.join (PYTHON_DIR, 'tcl', 'tcl8.6')
    os.environ ['TK_LIBRARY'] = os.path.join (PYTHON_DIR, 'tcl', 'tk8.6')
    include_files + = [
        (os.path.join (PYTHON_DIR, 'DLLs', 'tcl86t.dll'), ''),
        (os.path.join (PYTHON_DIR, 'DLLs', 'tk86t.dll'), '')
    ]

**Comment by ratansingh98** [comment]
> > Can you please provide a description of your error?
> 
> What I don't understand is that it searches for tensorflow in sites-application while I deposit it in the application folder

Yes, if you make an installer of any application it searches in sites-application. You can try **Hidden-import** parameter to explicitly link tensorflow to your build.

**Comment by Lucysmith8** [comment]
> Yes, if you make an installer of any application it searches in sites-application. You can try Hidden-import parameter to explicitly link tensorflow to your build.

How can I do this?

In my setup, I import tensorflow.

    options={
        'build_exe': {
            'packages': ['google', 'scipy', 'tensorflow'],

**Comment by ratansingh98** [comment]
Can you share your approach using Pyinstaller? I have a better understanding of making the installer in it.

**Comment by ratansingh98** [comment]
Can you share your error when you tried using Pyinstaller? I have a better understanding of making the installer in it.

**Comment by amahendrakar** [comment]
@Lucysmith8,
In order to expedite the trouble-shooting process, could you please provide the complete code to reproduce the issue reported here and the TensorFlow version you are using.

Also, please take a look at [this similar](https://github.com/pyinstaller/pyinstaller/issues/4200) issue and let us know if it works. Thanks!

**Comment by Lucysmith8** [comment]
> @Lucysmith8,
> In order to expedite the trouble-shooting process, could you please provide the complete code to reproduce the issue reported here and the TensorFlow version you are using.
> 
> Also, please take a look at [this similar](https://github.com/pyinstaller/pyinstaller/issues/4200) issue and let us know if it works. Thanks!

@amahendrakar I changed the question to add world code.

**Comment by Lucysmith8** [comment]
> Can you share your error when you tried using Pyinstaller? I have a better understanding of making the installer in it.

@ratansingh98 I changed the question to add world code.

**Comment by amahendrakar** [comment]
@Lucysmith8,
Please take a look at [this comment](https://github.com/pyinstaller/pyinstaller/issues/3362#issuecomment-375843871) from a similar issue and [this issue](https://github.com/pyinstaller/pyinstaller/issues/3942) thread, which seem to work.

Also, looks like this issue is a duplicate of [#40458](https://github.com/tensorflow/tensorflow/issues/40458). If it is could you please close one of them. Thanks!

**Comment by ratansingh98** [comment]
> > Can you share your error when you tried using Pyinstaller? I have a better understanding of making the installer in it.
> 
> @ratansingh98 I changed the question to add world code.

Thank you @Lucysmith8 , i will look into it.

**Comment by Lucysmith8** [comment]
He found the solution. But instead of sharing, he closed his question and he didn't answer anyone. https://github.com/anthony-tuininga/cx_Freeze/issues/514

**Comment by ratansingh98** [comment]
> > Can you share your error when you tried using Pyinstaller? I have a better understanding of making the installer in it.
> 
> @ratansingh98 I changed the question to add world code.

y

In the case of Pyinstaller, use TensorFlow 1.14 and add 'pkg_resources.py2_warn' as a hidden import.
